File: 04-april/08-xor-after-range-multiplication-queries-i/solution.java
class Solution {
    public int xorAfterQueries(int[] nums, int[][] queries) {
        final long MOD = 1_000_000_007L;
        long[] arr = new long[nums.length];
        for (int i = 0; i < nums.length; i++) arr[i] = nums[i];
        for (int[] q : queries) {
            int l = q[0], r = q[1], k = q[2];
            long v = q[3];
            for (int idx = l; idx <= r; idx += k) {
                arr[idx] = arr[idx] * v % MOD;
            }
        }
        long xor = 0;
        for (long x : arr) xor ^= x;
        return (int) xor;
    }
}

File: 04-april/08-xor-after-range-multiplication-queries-i/solution.js
var xorAfterQueries = function(nums, queries) {
    const MOD = 1_000_000_007n;
    const arr = nums.map(BigInt);
    for (const [l, r, k, v] of queries) {
        const bv = BigInt(v);
        for (let idx = l; idx <= r; idx += k) {
            arr[idx] = arr[idx] * bv % MOD;
        }
    }
    return arr.reduce((acc, x) => acc ^ x, 0n);
};
